### PR TITLE
Metal Pirates Room R-Mode Spark Interrupt

### PR DESCRIPTION
**Pull request by aquanight**
Room notes:
- Twelve drops from the metal pirates, roughly even split chance between Large Energy and Super. No Small or Nothing drop, so the farming's pretty good in this room. (Missile drop rate is below "worry about it" threshold.)
- Because of the amount of drops and the guarantee of Large Energy, 240 energy can be farmed.
- Three strats from each door:
  - Pirates Alive (CF only) strat is viable in randomizers that don't lock the door (e.g. MapRando)
  - Speed Echoes Kill is separate from regular kill to avoid hiding the notable. Because both pirates are being killed via shinesparks, the actual spark interrupt must happen through a heat interrupt.
  - Regular kill can start by killing one and hitbox the other to do a regular enemy interrupt, but otherwise it can do a heat interrupt.

Metal Pirates is also one of a few rooms where the blue suit does not itself assist with farming after gaining it. While it'll allow Samus to walk through the pirates and be invulnerable to their attacks while moving, it won't do any damage to them, even when they're in the vulnerable jump-kick animation. Generally, killing the pirate(s) after gaining blue suit should be treated like a normal kill.

Between this and Ridley Tank we now have two rooms with a strat that has a mandatory heat interrupt. This might be a good time to think about wrapping it in a helper?

### CODE REVIEW

File: region/lowernorfair/east/Metal Pirates Room.json [hunk 1]
```diff
@@ -253,6 +253,244 @@
...
+            {"or": [
+              {"and": [
+                {"resourceMissingAtMost": [{"type": "Super", "count": 0}]},
+                {"partialRefill": {"type": "ReserveEnergy", "limit": 240}}
+              ]},
+              {"partialRefill": {"type": "ReserveEnergy", "limit": 120}}
+            ]}
+          ]},
+          {"and": [
+            {"disableEquipment": "ETank"},
```
Comment (osse101) [inline region/lowernorfair/east/Metal Pirates Room.json]: Collecting drops from both pirates would take either
- Varia
- One round each pirate with supers or Plasma + X
- split your damage by switching which pirate you are damaging so they die at the same time?

File: region/lowernorfair/east/Metal Pirates Room.json [hunk 2]
```diff
@@ -253,6 +253,244 @@
...
+        ]},
+         "h_shinechargeMaxRunway",
+        {"disableEquipment": "Varia"},
+        {"disableEquipment": "Gravity"},
+        {"or": [
+          {"and": [
+            "canHitbox",
+            {"autoReserveTrigger": {"maxReserveEnergy": 95}},
+            "canRModeSparkInterrupt"
+          ]},
```
[the remark this answers is not in the document]
Comment (aquanight) [reply region/lowernorfair/east/Metal Pirates Room.json]: This branch represents killing one pirate, hitbox+runway and spark interrupt from second, then kill it after. (The other branches are heat Interrupt with both pirates dead.)

[the remark this answers is not in the document]
Comment (osse101) [reply region/lowernorfair/east/Metal Pirates Room.json]: tldr; I think this branch needs varia because of `{"maxReserveEnergy": 95}`

So we are into wacky item requirements where you can take 1000 heat frames on reserve->regular energy and then still have 99 you can disable down to.
And are doing spark interrupt strats without wanting to use speed echoes.

The 95 reserve energy limit means you can't safely kill the second pirate without varia.  And it would cause a problem if the obstacle being cleared unlocked a grey door you couldn't really use.  

And these higher heatframe counts might also actually require varia, but I think DASH provided items that reduced heat damage by 75% in certain regions, so who knows.

[the remark this answers is not in the document]
Comment (aquanight) [reply region/lowernorfair/east/Metal Pirates Room.json]: Might just have to do that, as I'm not sure how to better represent "splitting" the kills like that. I'd considered just having Speed Echoes be the way but after seeing that was a notable strat, it made me wary of making it the only farming option.

This branch does still allow a Crystal Flash which could also solve the energy issue, though it's logically being treated as happening after the pirate kill(s). Early CF is possible, but early CF with `canHitBox` (needed to use this branch) would just point you at the "Pirates Alive" strat which can then be followed by a normal kill the pirates (slightly safer for blue suit invulnerability, though slower for blue suit movement) to unlock the door.

I'll need to do some more testing of the ninja pirate behavior - perhaps there may be a way to just lift the 95 reserve maximum (provoke a jumpkick, then jump toward them as they are rising/kicking maybe?).

Another option I want to look at, is that I tested this idea on suitless 0+1:
- Fast single pirate kill (supers or microwave)
- Picking up the 120 energy that drops (possible that it doesn't even need a pause abuse)
- Hitbox through the remaining pirate.
- Run back toward the door you came from (the pirate should jump so you can often just run under him)
- You usually have time to open the door, either early or now as you're running toward it, if needed.
- Shine charge, get close to the transition for heat interrupt, then immediately get out.

Since it doesn't kill both pirates, this idea would only work if the door doesn't lock (MapRando BT-style door) or if it's already unlocked by killing the pirates once normally. (I think the Charge Plasma Double Hit Kill strat can kill fast enough for 0+1 suitless?)

[the remark this answers is not in the document]
Comment (aquanight) [reply region/lowernorfair/east/Metal Pirates Room.json]: Confirmed that a timed jump into the pirate kick attack can allow you to interrupt and the pirate ends up off-screen and won't bother you for a 400-point refill.
